File: python/datastructs/queue.py
from typing import Generic,TypeVar, override
T = TypeVar("T")

class Queue(Generic[T]):

    def __init__(self) -> None:
        self.size :int = 0;
        self.queue:list[T] = [];

    def enqueue(self,val:T) -> None:
        self.queue.append(val);
        self.size +=1;

    def dequeue(self) ->None:
        if self.isEmpty():
            return;
        _ = self.queue.pop(0);
        self.size -=1;
        
    def isEmpty(self)->bool:
        if len(self.queue) == 0:
            return True
        else:
            return False;

    def peek(self) ->str:
        if self.isEmpty():
            return "Queue is empty";
        return str(self.queue[0]);

    def len(self) ->int :
        return self.size;

    @override
    def __str__(self) -> str:
        string_to_return = "[";
        for item in self.queue:
            string_to_return += " " + str(item)  ;
        return string_to_return + " ]";


myQueue:Queue[int] = Queue();

myQueue.enqueue(8);
myQueue.enqueue(2);

myQueue.dequeue();

print(str(myQueue));

File: cpp/dataStructures/Queue.cpp
#include <iostream>
// WRITE YOUR CODE HERE
class Node {
  public:
  int data;
  Node* next;

  Node(int data){
    this->data = data;
    this->next = nullptr;
  }
};
class LinkedListQueue {
  private:
  Node* front;
  Node* back;

  public:
  LinkedListQueue() {
    front = nullptr;
    back = nullptr;
  }

  void enqueue(int value) {
    Node* newNode = new Node(value);
    if(front == nullptr){
      front = newNode;
      back = newNode;
    }
    else {
      back->next = newNode;
      back = newNode;
    }
  }
  int  dequeue() {
    if(isEmpty()){
      throw std::runtime_error("The Queue is empty " );
    }
    Node* temp = front;
    int frontData = temp->data;
    front = front->next;
    delete temp;
    if(isEmpty()){
      back = nullptr;
    }

    return frontData;
  }
  int peek() {
    if(isEmpty()){
      throw std::runtime_error("Queue is empty");
    }
    return front->data;
  }
  bool isEmpty() {
    return front == nullptr;
  }

  ~LinkedListQueue(){
    while(!isEmpty()){
      dequeue();
    }
  }

};

int main() {
    LinkedListQueue myQueue;
    myQueue.enqueue(1);
    myQueue.enqueue(2);
    myQueue.enqueue(3);
    std::cout << "Peek: " << myQueue.peek() << std::endl;
    std::cout << "Dequeue: " << myQueue.dequeue() << std::endl;
    std::cout << "Peek: " << myQueue.peek() << std::endl;

    return 0;
}
